resolwe/elastic/indices.py

```diff
@@ -0,0 +1,200 @@
+            raise RuntimeError('`document_class` must be defined in {}'.format(class_name))
+
+        self._index_name = self.document_class()._get_index()  # pylint: disable=not-callable,protected-access
+
+    def preprocess_object(self, obj):
+        """Preprocess object before indexing.
+
+        This function is called before `func:process_object`. It can be
+        used for advanced pre-processing of the object, i.e. adding
+        annotations that will be used in multiple fields.
+        """
+        pass
+
+    def get_object_id(self, obj):
+        """Return unique identifier of the object.
+
+        Object's id is returned by default. This method can be overriden
+        if object doesn't have ``id`` attribute.
+        """
+        return obj.id
+
+    def generate_id(self, obj):
+        """Generate unique document id for ElasticSearch."""
+        object_type = type(obj).__name__.lower()
+        return '{}_{}_{}'.format(object_type, self._index_name, self.get_object_id(obj))
+
+    def process_object(self, obj):
+        """Process current object and push it to the ElasticSearch."""
+        document = self.document_class(meta={'id': self.generate_id(obj)})  # pylint: disable=not-callable
+
+        for field in document._doc_type.mapping:  # pylint: disable=protected-access
+            if field in ['users_with_permissions', 'groups_with_permissions']:
+                continue  # this fields are handled separately
+
+            # use get_X_value function
+            get_value_function = getattr(self, 'get_{}_value'.format(field), None)
+            if get_value_function:
+                setattr(document, field, get_value_function())
+                continue
+
+            # use `mapping` dict
+            if field in self.mapping:
+                if callable(self.mapping[field]):
+                    setattr(document, field, self.mapping[field](obj))
+                    continue
+
+                object_attr = dict_dot(obj, self.mapping[field])
+                if callable(object_attr):
+                    # use method on object
+                    setattr(document, field, object_attr(obj))
+                else:
+                    # use attribute on object
+                    setattr(document, field, object_attr)
+                continue
+
+            # get value from the object
+            try:
+                object_value = dict_dot(obj, field)
+                setattr(document, field, object_value)
+                continue
+            except KeyError:
+                pass
+
+            raise AttributeError('Cannot determine mapping for field {}'.format(field))
+
+        permissions = self.get_permissions(obj)
+        document.users_with_permissions = permissions['users']
+        document.groups_with_permissions = permissions['groups']
+
+        document.save()
+
+    def run(self, obj=None):
+        """Main function for running indexes."""
+        if obj and obj not in self.queryset:
```

Comment:
Ok, if it is just for a single object (and never used in loops with a large number of objects) then it is ok.